File: src/content/3.6/code/scala/snippet21.scala
implicit def state[S] = new Adjunction[Prod[S, ?], Reader[S, ?]] {
  def counit[A](a: Prod[S, Reader[S, A]]): A = a match {
    case Prod((Reader(f), s)) => f(s)
  }

  def unit[A](a: A): Reader[S, Prod[S, A]] =
    Reader(s => Prod((a, s)))
}

File: src/content/3.6/code/ocaml/snippet21.ml
module AdjunctionState (S : sig
  type s
end) :
  Adjunction with type 'a f = (S.s, 'a) prod
  and type 'a r = (S.s, 'a) reader =
struct
  type 'a f = (S.s, 'a) prod
  type 'a r = (S.s, 'a) reader

  let counit (Prod (Reader f, s)) = f s
  let unit a = Reader (fun s -> Prod (a, s))
end
